hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/CallerContext.java

```diff
@@ -109,11 +113,32 @@ public String toString() {
 
   /** The caller context builder. */
   public static final class Builder {
-    private final String context;
+    private static final String KEY_VALUE_SEPARATOR = ":";
+    private final String fieldSeparator;
+    private final StringBuilder sb = new StringBuilder();
     private byte[] signature;
 
     public Builder(String context) {
-      this.context = context;
+      fieldSeparator = HADOOP_CALLER_CONTEXT_SEPARATOR_DEFAULT;
+      if (isValid(context)) {
+        sb.append(context);
+      }
+    }
+
+    public Builder(Configuration conf) {
+      fieldSeparator = conf.get(HADOOP_CALLER_CONTEXT_SEPARATOR_KEY,
+          HADOOP_CALLER_CONTEXT_SEPARATOR_DEFAULT);
+    }
```

Comment:
After adding this constructor, `new Builder(null)` fails. This is an incompatible change because CallerContext is `@Public`.

Reply:
Good catch.
Refactor builder constructor and remove unnecessary type case in test.
Please review, Thanks